File: blockchain/src/trade/trade.controller.ts
import {
  Controller,
  Get,
  Post,
  HttpException,
  HttpStatus,
  Param,
  Body,
} from '@nestjs/common';
import { TradeService } from './trade.service';
import { ResponseDTO } from '../dto/ResponseDTO';
import { BidDto } from '../dto/trade-bid-dto';

@Controller('trade')
export class TradeController {
  constructor(private tradeService: TradeService) {}

  @Get('contract-name')
  async getMarketPlaceName(): Promise<
    ResponseDTO<{ marketplaceName: string }>
  > {
    try {
      const marketplaceName = await this.tradeService.getMarketPlaceName();
      return new ResponseDTO('MarketPlace Name retrieved successfully', {
        marketplaceName,
      });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to retrieve MarketPlace Name', {
          details: error.message,
        }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Get('nft-active/:addressNFTCollection/:nftId')
  async getNFTActive(
    @Param('addressNFTCollection') addressNFTCollection: string,
    @Param('nftId') nftId: number,
  ): Promise<ResponseDTO<{ nftAllowedBid: boolean }>> {
    try {
      const nftAllowedBid = await this.tradeService.getNFTActive(
        addressNFTCollection,
        nftId,
      );
      return new ResponseDTO('NFT active status retrieved successfully', {
        nftAllowedBid,
      });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to retrieve NFT active status', {
          details: error.message,
        }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Get('nft-bids/:addressNFTCollection/:nftId')
  async getNFTBids(
    @Param('addressNFTCollection') addressNFTCollection: string,
    @Param('nftId') nftId: number,
  ): Promise<ResponseDTO<{ bidsOnNFT: BidDto[] }>> {
    try {
      const bidsOnNFT = await this.tradeService.getNFTBids(
        addressNFTCollection,
        nftId,
      );
      const mappedBids: BidDto[] = bidsOnNFT.map((bid) => ({
        addressNFTCollection: bid[0],
        nftId: bid[1],
        addressPaymentToken: bid[2],
        amountPaymentToken: bid[3],
        endTime: bid[4],
        bidder: bid[5],
      }));
      return new ResponseDTO('NFT bids retrieved successfully', {
        bidsOnNFT: mappedBids,
      });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to retrieve NFT bids', {
          details: error.message,
        }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Get('nft-bid-info/:addressNFTCollection/:nftId/:bidId')
  async getNFTBidInfo(
    @Param('addressNFTCollection') addressNFTCollection: string,
    @Param('nftId') nftId: number,
    @Param('bidId') bidId: number,
  ): Promise<ResponseDTO<{ bidInfo: BidDto }>> {
    try {
      const bidInfo = await this.tradeService.getNFTBidInfo(
        addressNFTCollection,
        nftId,
        bidId,
      );
      const mappedBidInfo: BidDto = {
        addressNFTCollection: bidInfo[0],
        nftId: bidInfo[1],
        addressPaymentToken: bidInfo[2],
        amountPaymentToken: bidInfo[3],
        endTime: bidInfo[4],
        bidder: bidInfo[5],
      };
      return new ResponseDTO('NFT bid info retrieved successfully', {
        bidInfo: mappedBidInfo,
      });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to retrieve NFT bid info', {
          details: error.message,
        }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Post('activate-bidding/:collection/:nftId')
  async activateBidding(
    @Param('collection') collection: string,
    @Param('nftId') nftId: number,
    @Body('privateKey') privateKey: string,
  ): Promise<ResponseDTO<{ txResult: any }>> {
    try {
      const txResult = await this.tradeService.activateBidding(
        collection,
        nftId,
        privateKey,
      );
      return new ResponseDTO('Bidding activated successfully', { txResult });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to activate bidding', {
          details: error.message,
        }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Post('deactivate-bidding/:collection/:nftId')
  async deactivateBidding(
    @Param('collection') collection: string,
    @Param('nftId') nftId: number,
    @Body('privateKey') privateKey: string,
  ): Promise<ResponseDTO<{ txResult: any }>> {
    try {
      const txResult = await this.tradeService.deactivateBidding(
        collection,
        nftId,
        privateKey,
      );
      return new ResponseDTO('Bidding de-activated successfully', { txResult });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to deactivate bidding', {
          details: error.message,
        }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Post('bid')
  async createBid(
    @Body('bidRequest') bidRequest: BidDto,
    @Body('privateKey') privateKey: string,
  ): Promise<ResponseDTO<{ txResult: any }>> {
    try {
      const txResult = await this.tradeService.createBid(
        bidRequest,
        privateKey,
      );
      return new ResponseDTO('Bid created successfully', { txResult });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to create bidding', { details: error.message }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Post('acceptBid/:collection/:nftId/:bidId')
  async acceptBid(
    @Param('collection') collection: string,
    @Param('nftId') nftId: number,
    @Param('bidId') bidId: number,
    @Body('privateKey') privateKey: string,
  ): Promise<ResponseDTO<{ txResult: any }>> {
    try {
      const txResult = await this.tradeService.acceptBid(
        collection,
        nftId,
        bidId,
        privateKey,
      );
      return new ResponseDTO('Bid accepted successfully', { txResult });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to accept bidding', { details: error.message }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }
}

File: spring-boot-server/src/main/java/org/example/springbootserver/trade/controller/TradeController.java
package org.example.springbootserver.trade.controller;

import lombok.RequiredArgsConstructor;
import org.example.springbootserver.global.dto.HttpResponseDTO;
import org.example.springbootserver.global.dto.HttpResponseDTOv2;
import org.example.springbootserver.trade.dto.BidDTO;
import org.example.springbootserver.trade.dto.BidRequestDTO;
import org.example.springbootserver.trade.dto.BidResponseDTO;
import org.example.springbootserver.trade.service.TradeService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/trade")
@RequiredArgsConstructor
public class TradeController {

    private static final Logger logger = LoggerFactory.getLogger(TradeController.class);
    private final TradeService tradeService;

    @GetMapping("/contract-name")
    public ResponseEntity<HttpResponseDTOv2> getMarketPlaceName() {
        return tradeService.getMarketPlaceName();
    }

    @GetMapping("/nft-active/{collection}/{nftId}")
    public  ResponseEntity<HttpResponseDTOv2>  getNFTActive(
            @PathVariable String collection,
            @PathVariable int nftId) {
        return tradeService.getNFTActive(collection, nftId);
    }

    @GetMapping("/nft-bids/{collection}/{nftId}")
    public ResponseEntity<HttpResponseDTOv2<Map<String, List<BidDTO>>>> getNFTBids(
            @PathVariable String collection,
            @PathVariable int nftId) {
        long startTime = System.currentTimeMillis();
        logger.info("Starting getNFTBids request for collection: {}, nftId: {}", collection, nftId);

        ResponseEntity<HttpResponseDTOv2<Map<String, List<BidDTO>>>> response = tradeService.getNFTBids(collection, nftId);

        logger.info("Received response from tradeService after {} ms", System.currentTimeMillis() - startTime);
        logger.debug("Response body: {}", response.getBody());

        long endTime = System.currentTimeMillis();
        logger.info("Completed getNFTBids request in {} ms", endTime - startTime);

        return response;
    }

//    @GetMapping("/nft-bids/{collection}/{nftId}")
//    public ResponseEntity<HttpResponseDTO<Map<String, List<BidDTO>>>> getNFTBids(
//            @PathVariable String collection,
//            @PathVariable int nftId) {
//        HttpResponseDTO<Map<String, List<BidDTO>>> bids = tradeService.getNFTBids(collection, nftId);
//        return new ResponseEntity<>(bids, HttpStatus.OK);
//    }

    @GetMapping("/nft-bid-info/{collection}/{nftId}/{bidId}")
    public ResponseEntity<HttpResponseDTOv2> getNFTBidInfo(
            @PathVariable String collection,
            @PathVariable int nftId,
            @PathVariable int bidId) {
        return tradeService.getNFTBidInfo(collection, nftId, bidId);
    }

    @PostMapping("/activateBidding/{nftId}")
    public ResponseEntity<HttpResponseDTOv2> activateBidding(@PathVariable int nftId) {
        return tradeService.activateBidding(nftId);
    }

    @PostMapping("/deactivateBidding/{nftId}")
    public ResponseEntity<HttpResponseDTOv2> deactivateBidding(@PathVariable int nftId) {
        return tradeService.deactivateBidding(nftId);
    }

    @PostMapping("/bid")
    public ResponseEntity<HttpResponseDTOv2> createBid(@RequestBody BidRequestDTO bidRequest) {
        return tradeService.createBid(bidRequest);
    }

    @PostMapping("/acceptBid/{nftId}/{bidId}")
    public ResponseEntity<HttpResponseDTOv2> acceptBid(
            @PathVariable int nftId,
            @PathVariable int bidId) {
        return tradeService.acceptBid(nftId, bidId);
    }

    // Get all bids for an NFT owned by Current User
    @GetMapping("/nft-bids/received")
    public ResponseEntity<HttpResponseDTOv2<List<BidResponseDTO>>> getAllBidsReceived() {
        List<BidResponseDTO> bids = tradeService.getAllBidsReceived();
        HttpResponseDTOv2<List<BidResponseDTO>> response = new HttpResponseDTOv2<>(
                "NFT bids Received retrieved successfully",
                bids
        );
        return new ResponseEntity<>(response, HttpStatus.OK);
    }

    @GetMapping("/nft-bids/proposed")
    public ResponseEntity<HttpResponseDTOv2<List<BidResponseDTO>>> getAllBidsProposedByUser() {
        List<BidResponseDTO> bids = tradeService.getAllBidsProposedByUser();
        HttpResponseDTOv2<List<BidResponseDTO>> response = new HttpResponseDTOv2<>(
                "NFT bids Proposed retrieved successfully",
                bids
        );
        return new ResponseEntity<>(response, HttpStatus.OK);
    }

    @GetMapping("/nft-bids/received_deprecated")
    public ResponseEntity<HttpResponseDTO<Map<String, List<BidResponseDTO>>>> getAllBidsReceivedDeprecated() {
        HttpResponseDTO<Map<String, List<BidResponseDTO>>> bids = tradeService.getAllBidsReceivedDeprecated();
        return new ResponseEntity<>(bids, HttpStatus.OK);
    }

    @GetMapping("/nft-bids/proposed_deprecated")
    public ResponseEntity<HttpResponseDTO<Map<String, List<BidResponseDTO>>>> getAllBidsProposedByUserDeprecated() {
        HttpResponseDTO<Map<String, List<BidResponseDTO>>> bids = tradeService.getAllBidsProposedByUserDeprecated();
        return new ResponseEntity<>(bids, HttpStatus.OK);
    }
}